### ISSUE TITLE
Platform selection: having evdi available stops Mir-on-X11 working

### ISSUE THREAD

**Issue by AlanGriffiths** [canonical/mir#3139]
This doesn't happen when Mir is the host, but does happen on Mutter: the failure mode is:

```
[2023-11-17 14:09:43.563204] <information> mirserver: Selected rendering driver: mir:egl-generic (version 2.15.0) for platform
[2023-11-17 14:09:43.579928] <information> mirserver: Selected rendering driver: mir:gbm-kms (version 2.15.0) for device ((null): /dev/dri/renderD128)
...
ERROR: /home/alan/CLionProjects/mir/src/server/compositor/default_display_buffer_compositor_factory.cpp(78): Throw in function virtual std::unique_ptr<mc::DisplayBufferCompositor> mir::compositor::DefaultDisplayBufferCompositorFactory::create_compositor_for(mg::DisplaySink &)
Dynamic exception type: boost::wrapexcept<std::logic_error>
std::exception::what: No rendering platform claims to support this output
```

To run Mir-on-X11 needs `--driver-quirks=skip:driver:evdi` or `--platform-rendering-libs=mir:egl-generic`

**Comment by AlanGriffiths** [comment]
So, the problem in this case is that evdi outouts are supported by gbm-kms but with display capability only, with no rendering.

The problem arises because we try `mir:gbm-kms` and fail; `--platform-rendering-libs=mir:egl-generic` is another way to get a working rendering driver.

**Comment by AlanGriffiths** [comment]
Another data point: when the host is Mir (Miriway), the `mir:gbm-kms` platform is **not** selected for rendering and "Mir-on-X11" works. When the host is Mutter (GNOME) the `mir:gbm-kms` platform is selected for rendering and "Mir-on-X11" fails.

**Comment by RAOF** [comment]
> So, the problem in this case is that evdi outouts are supported by gbm-kms but with display capability only, with no rendering.

Actually, I'm fairly certain the problem here is that the `gbm-kms` rendering platform cannot output *to X11*, and if probing had been the other way around and `egl-generic` had been selected as `the_buffer_allocator` then this would all have worked.

Relatedly, it might be nice if
> std::exception::what: No rendering platform claims to support this output

told us *which* output “this” is.

> When the host is Mutter (GNOME) the mir:gbm-kms platform is selected for rendering and "Mir-on-X11" fails

It looks to me like Mutter drops DRM master on cards which have no outputs it's driving¹ (and, presumably, which it's not rendering on). This means that, under Mutter, the `gbm-kms` *display* platform is loadable on the evdi output, which means the `gbm-kms` *rendering* platform is loadable (because it doesn't require exclusive access to the hardware).

¹: There's a pretty good reason for this; if nothing has an open handle to the DRM device, you can happily power the whole thing off

**Comment by Saviq** [comment]
> ¹: There's a pretty good reason for this; if nothing has an open handle to the DRM device, you can happily power the whole thing off

What if you then connect an output, though?

**Comment by RAOF** [comment]
That's a fair question! I'd the card isn't like PCIe-powered-off, I think it'll still generate a udev hotplug event?

**Comment by AlanGriffiths** [comment]
This is fixed in 2.17, I've not tested 2.16
